### PR TITLE
Adguard dns analyzer, closes #1361

### LINKED ISSUE #1361: [Analyzer] AdGuard DNS
## Name
AdGuard

## Link
https://adguard-dns.io/kb/it/general/dns-providers/

## Type of analyzer
this should be used as domain/url analyzer

## Why should we use it
We can get DNS resolution + whether the analyzed observable is something that the ADGuard service would block or not

## Possible implementation
First analyzer would just be classic dns resolution, without filters: https://unfiltered.adguard-dns.com/dns-query
Second analyzer would be used to detect malicious behavior, use https://dns.adguard-dns.com/dns-query and check for null responses

### PR DESCRIPTION
**Pull request by g4ze**
close #1361 

# Description

Please include a summary of the change and link to the related issue.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [x] A new plugin (analyzer, connector, visualizer, playbook, pivot or ingestor) was added or changed, in which case:
    - [x] I strictly followed the documentation ["How to create a Plugin"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-add-a-new-plugin)
    - [x] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [x] I have dumped the configuration from Django Admin using the `dumpplugin` command and added it in the project as a data migration. (["How to share a plugin with the community"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-share-your-plugin-with-the-community))
    - [ ] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [x] If you created a new analyzer and it is free (does not require any API key), please add it in the `FREE_TO_USE_ANALYZERS` playbook by following [this guide](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-modify-a-plugin).
    - [x] Check if it could make sense to add that analyzer/connector to other [freely available playbooks](https://intelowl.readthedocs.io/en/develop/Usage.html#list-of-pre-built-playbooks).
    - [x] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
    - [x] If the plugin interacts with an external service, I have created an attribute called precisely `url` that contains this information. This is required for Health Checks. 
    - [ ] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [ ] I have added that raw JSON sample to the `MockUpResponse` of the `_monkeypatch()` method. This serves us to provide a valid sample for testing.
- [ ] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Contribute.md)).
- [ ] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.
- [ ] After you had submitted the PR, if `DeepSource`, `Django Doctors` or other third-party linters have triggered any alerts during the CI checks, I have solved those alerts.

### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.

image

### CODE REVIEW

**Comment by g4ze** [comment]
@mlodic  I am not able to figure out the right query for the url. seems like it expects dns in an encoded form. Can you please take a look into it?

**Comment by mlodic** [comment]
following the DMs we had, there are some pointers that can help solving this issue:
* https://github.com/AdguardTeam/AdGuardDNS/issues/26 or maybe other issues talking about them
* https://datatracker.ietf.org/doc/html/rfc8484 -> official format for dns over https queries, they could follow something written on this.

If you could try to do some more advanced research about that, that would be cool.

**Comment by g4ze** [comment]
figured out a way, will complete this PR this week :)

**Comment by g4ze** [comment]
I think the codefactor ci test is giving a fall positive to the sinkhole IP. Can you please have a look at it?

**Comment by g4ze** [comment]
kindly check the TLP and review.

**Review by drosetti — CHANGES_REQUESTED** [review]

File: api_app/analyzers_manager/observable_analyzers/dns/dns_malicious_detectors/adguard.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    @staticmethod
+    def check_a(observable: str, a_filtered: List[RRset]) -> dict:
+        # adguard follows 2 patterns for malicious domains,
+        # it either redirects the request to ad-block.dns.adguard.com
+        # or it sinkholes the request (to 0.0.0.0).
+        # If the response contains neither of these,
+        # we can safely say the domain is not malicious
+        for ans in a_filtered:
+            if str(ans.name) == "ad-block.dns.adguard.com.":
```
[the remark this answers is not in the document]
Comment (g4ze) [reply api_app/analyzers_manager/observable_analyzers/dns/dns_malicious_detectors/adguard.py:59]: Yes, I did that after checking it with the response body. Plus it works as expected upon testing.

File: api_app/analyzers_manager/observable_analyzers/dns/dns_malicious_detectors/adguard.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                return malicious_detector_response(
+                    observable=observable, malicious=True
+                )
+
+            if any(str(data) == "0.0.0.0" for data in ans):
```
[the remark this answers is not in the document]
Comment (g4ze) [reply api_app/analyzers_manager/observable_analyzers/dns/dns_malicious_detectors/adguard.py:64]: i think [this](https://support.codefactor.io/i15-ignoring-code-files) is something only maintainers can do.

**Review by mlodic — APPROVED** [review]
really appreciated the way you moved to extract all the information required to solve the problem. The PR seems nice, I'll merge it.

**Comment by g4ze** [comment]
All thanks to you and Daniele🙏

File: api_app/analyzers_manager/observable_analyzers/dns/dns_malicious_detectors/adguard.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        return dns.message.from_wire(r_filtered.content).answer
+
+    @staticmethod
+    def check_a(observable: str, a_filtered: List[RRset]) -> dict:
+        # adguard follows 2 patterns for malicious domains,
+        # it either redirects the request to ad-block.dns.adguard.com
+        # or it sinkholes the request (to 0.0.0.0).
+        # If the response contains neither of these,
+        # we can safely say the domain is not malicious
+        for ans in a_filtered:
```
[the remark this answers is not in the document]
Comment (g4ze) [reply api_app/analyzers_manager/observable_analyzers/dns/dns_malicious_detectors/adguard.py:58]: Personal tries
Queried both the links with good and bad domains.

File: api_app/playbooks_manager/migrations/0049_add_adguard_to_free_to_use.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("playbooks_manager", "0048_playbook_config_download_file"),
+        ("analyzers_manager", "0101_analyzer_config_adguard"),
+    ]
+
+    operations = [
+        migrations.RunPython(migrate, reverse_migrate),
+    ]
```
Comment (mlodic) [inline api_app/playbooks_manager/migrations/0049_add_adguard_to_free_to_use.py:34]: ah I forgot to ask you another thing: can you add it in the `Dns` default playbook?

Comment (g4ze) [reply api_app/playbooks_manager/migrations/0049_add_adguard_to_free_to_use.py:34]: Yes, will do that in a while!

**Comment by g4ze** [comment]
should be merged first acc to [this](https://github.com/intelowlproject/IntelOwl/pull/2402#issuecomment-2202346641)

**Pull request merged**